gitpod-io/gitpod
# [JetBrains EAP] Port Forwarding improvements

## Description

This PR brings the following changes:

- We now rely on `PerClientPortForwardingService` for controlling all Ports (Forwarded + Exposed) in JetBrains Client. For this reason, `GitpodPortsService` was removed.
- `GitpodCLIService.kt` and `GitpodClientProjectSessionTracker.kt` were updated because they relied on `GitpodPortsService`.
- Whenever Supervisor informs the client there was a change in the Name, Description, Visibility, Served or Exposed status of a port, Ports View is updated accordingly.
- `GitpodPortForwardingService` became an interface, which is implemented only by the Latest package. It's not available for Stable.
- `extractLocalHostUriMetaDataForPortMapping` moved to an object (`LocalHostUri`) as it's a pure function with its test suit. Leaving it inside a Service would make it hard to instantiate and test.
- `GitpodGlobalPortForwardingService` was removed as we shouldn't use JetBrains' GlobalPortForwardingService. This is intended for plugins running in local mode instead of remote.

This is how the Ports View currently looks like:
<img width="696" alt="image" src="https://user-images.githubusercontent.com/418083/201307506-e0d86c2d-6a8d-42b6-a9e6-85844dd4c11e.png">

Notes:
- By default, when trying to forward a port which is already in use on Client, a random port will be forwarded (and we take it into account when fetching the forwarded port on client).
  <img width="437" alt="image" src="https://user-images.githubusercontent.com/418083/200579227-e9669d7c-f3b3-4458-ad82-5f72602300d1.png">
  ![image](https://user-images.githubusercontent.com/418083/201308185-d3b17cb0-fd02-4859-9dc2-09edfad0ea22.png)
- `GitpodPortForwardingService` is only available when using Latest (EAP) IDEs, that's why we use `serviceOrNull<GitpodPortForwardingService>()` when fetching that service.
- Currently, there's no way to sort the ports in Ports View. They appear in the same order they're registered.

## Related Issue(s)

- Follow-up of https://github.com/gitpod-io/gitpod/pull/14356
- Affected by https://github.com/gitpod-io/gitpod/issues/11849 [See [this comment](https://github.com/gitpod-io/gitpod/pull/14408#discussion_r1016610904)]

## How to test

1. Open the Preview Environment of this PR and start any workspace using _EAP IntelliJ IDEA_.
  - Run `python -m http.server 3000` and confirm that:
    - a notification is displayed informing that the port is available. Then click "Open browser" and confirm it opens <http://127.0.0.1:3000>.
    - port 3000 is displayed on the Ports View twice: one with the forwarded address and another with the exposed URL.
    - `Copy Web URL` and `Copy URL` dropdown options are working:
      - ![image](https://user-images.githubusercontent.com/418083/200541869-950a7614-75e9-4a52-a31e-b6f84c9c2dbc.png)
      - ![image](https://user-images.githubusercontent.com/418083/200541894-7280bf05-c60e-4e01-a932-1ca28bc9f0ed.png)
  - In another terminal run `gp ports visibility 3000:public` and confirm if in Ports View the Exposed Port doesn't have the lock icon anymore.
2. Open the Preview Environment of this PR and start any workspace using _Stable IntelliJ IDEA_.
  - Run `python -m http.server 3000` and confirm that a notification appears with a link to "Open browser". Click the link and check if it opens a Gitpod URL, as there's no port forwarding in the Stable version.

## Release Notes

```release-note
NONE
```

## Werft options:

      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`
